process/smartContract/process.go

```diff
@@ -464,14 +464,27 @@ func (sc *scProcessor) isSelfShard(address []byte) bool {
 	return sc.shardCoordinator.ComputeId(address) == sc.shardCoordinator.SelfId()
 }
 
+func safeSubUint64(a, b uint64) uint64 {
+	if a < b {
+		return 0
+	}
+	return a - b
+}
+
 func (sc *scProcessor) computeTotalConsumedFeeAndDevRwd(
 	tx data.TransactionHandler,
 	vmOutput *vmcommon.VMOutput,
 	builtInFuncGasUsed uint64,
 ) (*big.Int, *big.Int) {
-	consumedGas := tx.GetGasLimit() - vmOutput.GasRemaining
-	if !sc.isSelfShard(tx.GetSndAddr()) {
-		consumedGas -= sc.economicsFee.ComputeGasLimit(tx)
+	if tx.GetGasLimit() == 0 {
+		return big.NewInt(0), big.NewInt(0)
+	}
+
+	senderInAnotherShard := !sc.isSelfShard(tx.GetSndAddr())
+	consumedGas := safeSubUint64(tx.GetGasLimit(), vmOutput.GasRemaining)
+	if senderInAnotherShard {
+		consumedGas = safeSubUint64(consumedGas, sc.economicsFee.ComputeGasLimit(tx))
+		consumedGas = safeSubUint64(consumedGas, builtInFuncGasUsed)
```

Comment:
It is correct to do this sub here? Now the totalFee from line 500 will be the same with totalFeeMinusBuiltIn from line 505 if senderInAnotherShard is true. Until now the functionality was different. totalFee was always the fee without the builtInFuncGasUsed substracted

Reply:
it is correct. as all gas in case of builtin is consumed in sender shard only.